File: src/KRATE.Core/Tools/Escapes.cs
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace Krate.Core;

public static class Escapes
{
    // JsonEncodedText rather than the serializer: no reflection, no source-gen context to maintain.
    public static string Json(string s) => '"' + JsonEncodedText.Encode(s).ToString() + '"';

    public static string JsonUnescape(string s)
    {
        var t = s.Trim();
        if (!t.StartsWith('"')) t = '"' + t + '"';
        using var doc = JsonDocument.Parse(t);
        return doc.RootElement.GetString() ?? "";
    }

    /// <summary>SQL string literal: doubles the quotes. Escaping is not a substitute for
    /// parameterised queries — this is for pasting a value into a console.</summary>
    public static string Sql(string s) => "'" + s.Replace("'", "''") + "'";

    /// <summary>POSIX shell literal: single quotes, with the standard '"'"' dance inside.</summary>
    public static string Shell(string s) => "'" + s.Replace("'", "'\"'\"'") + "'";

    /// <summary>Decodes a JWT locally: header and payload, dates spelled out.
    /// The signature is NOT verified — that needs the issuer's key.</summary>
    public static string Jwt(string input)
    {
        var parts = input.Trim().Split('.');
        if (parts.Length < 2) throw new ArgumentException(Strings.Get("Error_BadJwt"));

        var header = Krate.Core.Json.Format(FromBase64Url(parts[0]));
        var payload = FromBase64Url(parts[1]);
        var claims = Krate.Core.Json.Format(payload);

        var dates = new List<string>();
        using (var doc = JsonDocument.Parse(payload))
            foreach (var claim in new[] { "iat", "nbf", "exp" })
                if (doc.RootElement.TryGetProperty(claim, out var v) && v.TryGetInt64(out var seconds))
                {
                    var when = DateTimeOffset.FromUnixTimeSeconds(seconds);
                    var note = claim == "exp" ? Strings.Get(when < DateTimeOffset.UtcNow ? "Jwt_Expired" : "Jwt_Valid") : "";
                    dates.Add($"{claim}  {when.UtcDateTime:yyyy-MM-dd HH:mm:ss}Z  {note}".TrimEnd());
                }

        var lines = new List<string> { Strings.Get("Jwt_Header"), header, "", Strings.Get("Jwt_Payload"), claims };
        if (dates.Count > 0) lines.AddRange(["", Strings.Get("Jwt_Dates"), .. dates]);
        lines.AddRange(["", Strings.Get("Jwt_NotVerified")]);
        return string.Join('\n', lines);
    }

    /// <summary>Convert.FromBase64String throws a raw, untranslated FormatException — a malformed
    /// token is a bad token, and the caller already has a message for that.</summary>
    static string FromBase64Url(string s)
    {
        var padded = s.Replace('-', '+').Replace('_', '/').PadRight((s.Length + 3) / 4 * 4, '=');
        try { return Encoding.UTF8.GetString(Convert.FromBase64String(padded)); }
        catch (FormatException) { throw new ArgumentException(Strings.Get("Error_BadJwt")); }
    }

    /// <summary>Plain and scientific notation of the same number, both directions.</summary>
    public static string Scientific(string input)
    {
        if (!double.TryParse(input.Trim().Replace(" ", ""), NumberStyles.Float, CultureInfo.InvariantCulture, out var value))
            throw new ArgumentException(Strings.Get("Error_NeedNumber"));
        return string.Join('\n',
            string.Create(CultureInfo.InvariantCulture, $"PLAIN       {value:0.##############################}"),
            string.Create(CultureInfo.InvariantCulture, $"SCIENTIFIC  {value:0.##############E+0}"),
            string.Create(CultureInfo.InvariantCulture, $"ENGINEERING {Engineering(value)}"));
    }

    static string Engineering(double v)
    {
        if (v == 0) return "0e0";
        var exponent = (int)Math.Floor(Math.Log10(Math.Abs(v)) / 3) * 3; // exponent is a multiple of 3
        return string.Create(CultureInfo.InvariantCulture, $"{v / Math.Pow(10, exponent):0.##########}e{exponent}");
    }

    public static string ToCrlf(string s) => s.Replace("\r\n", "\n").Replace("\n", "\r\n");
    public static string ToLf(string s) => s.Replace("\r\n", "\n");

    /// <summary>Windows ↔ Unix path separators, direction detected from the input.</summary>
    public static string Path(string s)
    {
        var t = s.Trim();
        return t.Contains('\\') ? t.Replace('\\', '/') : t.Replace('/', '\\');
    }

    /// <summary>Makes a filename Windows-safe: forbidden characters, trailing dots and
    /// reserved device names (CON, NUL…) all handled.</summary>
    public static string Filename(string s)
    {
        var name = new string(s.Trim().Select(c => System.IO.Path.GetInvalidFileNameChars().Contains(c) ? '_' : c).ToArray())
            .TrimEnd(' ', '.');
        if (name.Length == 0) return "_";
        var stem = System.IO.Path.GetFileNameWithoutExtension(name);
        string[] reserved = ["CON", "PRN", "AUX", "NUL", .. Enumerable.Range(1, 9).SelectMany(i => new[] { $"COM{i}", $"LPT{i}" })];
        return reserved.Contains(stem, StringComparer.OrdinalIgnoreCase) ? "_" + name : name;
    }
}

File: rust/src/escapes.rs
//! Quoting a value for another language's syntax. Mirrors `Krate.Core.Escapes`.

/// JSON string literal, quotes included.
pub fn json(s: &str) -> String {
    let mut out = String::with_capacity(s.len() + 2);
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\u0022"),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{8}' => out.push_str("\\b"),
            '\u{c}' => out.push_str("\\f"),
            // JsonEncodedText escapes the HTML-sensitive characters too, so the result is safe
            // to drop straight into a <script> block.
            '<' => out.push_str("\\u003C"),
            '>' => out.push_str("\\u003E"),
            '&' => out.push_str("\\u0026"),
            '\'' => out.push_str("\\u0027"),
            '+' => out.push_str("\\u002B"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04X}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// SQL string literal: doubles the quotes. Escaping is not a substitute for parameterised
/// queries — this is for pasting a value into a console.
pub fn sql(s: &str) -> String {
    format!("'{}'", s.replace('\'', "''"))
}

/// POSIX shell literal: single quotes, with the standard '"'"' dance inside.
pub fn shell(s: &str) -> String {
    format!("'{}'", s.replace('\'', "'\"'\"'"))
}

/// Windows and Unix path separators, direction detected from the input.
pub fn path(s: &str) -> String {
    let t = s.trim();
    if t.contains('\\') { t.replace('\\', "/") } else { t.replace('/', "\\") }
}

const INVALID_FILENAME: [char; 9] = ['"', '<', '>', '|', ':', '*', '?', '\\', '/'];

/// Makes a filename Windows-safe: forbidden characters, trailing dots and reserved device
/// names (CON, NUL…) all handled.
pub fn filename(s: &str) -> String {
    let cleaned: String = s
        .trim()
        .chars()
        .map(|c| if INVALID_FILENAME.contains(&c) || (c as u32) < 0x20 { '_' } else { c })
        .collect();
    let name = cleaned.trim_end_matches([' ', '.']);
    if name.is_empty() {
        return "_".to_string();
    }

    let stem = name.rsplit_once('.').map_or(name, |(head, _)| head);
    let mut reserved = vec!["CON".to_string(), "PRN".to_string(), "AUX".to_string(), "NUL".to_string()];
    for i in 1..=9 {
        reserved.push(format!("COM{i}"));
        reserved.push(format!("LPT{i}"));
    }
    if reserved.iter().any(|r| r.eq_ignore_ascii_case(stem)) {
        format!("_{name}")
    } else {
        name.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn json_quotes_and_escapes() {
        assert_eq!(json("hi"), "\"hi\"");
        assert_eq!(json("a\nb"), "\"a\\nb\"");
        assert_eq!(json("tab\there"), "\"tab\\there\"");
        assert!(json("</script>").contains("\\u003C"), "HTML-sensitive chars are escaped");
    }

    #[test]
    fn sql_doubles_quotes() {
        assert_eq!(sql("O'Brien"), "'O''Brien'");
        assert_eq!(sql("plain"), "'plain'");
    }

    #[test]
    fn shell_survives_embedded_quotes() {
        assert_eq!(shell("plain"), "'plain'");
        assert_eq!(shell("it's"), "'it'\"'\"'s'");
    }

    #[test]
    fn path_flips_whichever_way_it_came() {
        assert_eq!(path(r"C:\a\b"), "C:/a/b");
        assert_eq!(path("C:/a/b"), r"C:\a\b");
    }

    #[test]
    fn filename_strips_what_windows_rejects() {
        assert_eq!(filename("a:b*c?.txt"), "a_b_c_.txt");
        assert_eq!(filename("trailing..."), "trailing");
        assert_eq!(filename("   "), "_");
        assert_eq!(filename("CON"), "_CON", "reserved device name");
        assert_eq!(filename("con.txt"), "_con.txt", "reserved even with an extension");
        assert_eq!(filename("normal.txt"), "normal.txt");
    }
}
